src/persist-client/src/read.rs

```diff
@@ -232,17 +298,48 @@ where
     ///
     /// If you have a use for consolidated listen output, given that snapshots can't be
     /// consolidated, come talk to us!
-    #[instrument(level = "debug", name = "listen::next", skip_all, fields(shard = %self.machine.shard_id()))]
+    #[instrument(level = "debug", name = "listen::next", skip_all, fields(shard = %self.handle.machine.shard_id()))]
     pub async fn next(&mut self) -> Vec<ListenEvent<K, V, T, D>> {
         trace!("Listen::next");
 
-        let batch = self.machine.next_listen_batch(&self.frontier).await;
+        // We might also want to push maybe_heartbeat into the
+        // `next_listen_batch` loop so that we can heartbeat even when batches
+        // aren't incoming. At the moment the ownership would be weird and we
+        // should at least be regularly getting frontier updates, so maybe it
+        // isn't actually necessary.
```

Comment:
> we should at least be regularly getting frontier updates

meaning that, even without incoming batches, we'll still be heartbeating regularly?